src/paginate.ts

```diff
@@ -175,6 +178,144 @@ export const DEFAULT_MAX_LIMIT = 100
+        return false
+    }
+    return !!qb?.expressionMap?.mainAlias?.metadata?.hasRelationWithPropertyPath(propertyPath)
+}
+
+// This function is used to fix the query parameters when using relation, embeded or virtual properties
+// It will replace the column name with the alias name and return the new parameters
+function fixQueryParam(
+    alias: string,
+    column: string,
+    filter: Filter,
+    condition: WherePredicateOperator,
+    parameters: { [key: string]: string }
+): { [key: string]: string } {
+    const isNotOperator = (condition.operator as string) === 'not'
+
+    const conditionFixer = (
+        alias: string,
+        column: string,
+        filter: Filter,
+        operator: WherePredicateOperator['operator'],
+        parameters: { [key: string]: string }
+    ): { condition_params: any; params: any } => {
+        let condition_params: any = undefined
+        let params = parameters
+        switch (operator) {
+            case 'between':
+                condition_params = [alias, `:${column}_from`, `:${column}_to`]
+                params = {
+                    [column + '_from']: filter[column].value[0],
+                    [column + '_to']: filter[column].value[1],
+                }
+                break
+            case 'in':
+                condition_params = [alias, `:...${column}`]
+                break
+            default:
+                condition_params = [alias, `:${column}`]
+                break
+        }
+        return { condition_params, params }
+    }
+
+    const { condition_params, params } = conditionFixer(
+        alias,
+        column,
+        filter,
+        isNotOperator ? condition['condition']['operator'] : condition.operator,
+        parameters
+    )
+
+    if (isNotOperator) {
+        condition['condition']['parameters'] = condition_params
+    } else {
+        condition.parameters = condition_params
+    }
+
+    return params
+}
+
+// This function is used to fix the column alias when using relation, embedded or virtual properties
+function fixColumnAlias(
+    properties: ColumnProperties,
+    alias: string,
+    isRelation = false,
+    isVirtualProperty = false,
+    query?: ColumnMetadata['query']
+): string {
+    if (isRelation) {
+        if (isVirtualProperty && query) {
+            return `(${query(`${alias}_${properties.propertyPath}`)})` // () is needed to avoid parameter conflict
+        } else if (isVirtualProperty && !query) {
+            return `${alias}_${properties.propertyPath}_${properties.propertyName}`
+        } else {
+            return `${alias}_${properties.propertyPath}.${properties.propertyName}` // include embeded property and relation property
+        }
+    } else if (isVirtualProperty) {
+        return query ? `(${query(`${alias}`)})` : `${alias}_${properties.propertyName}`
+    } else {
+        return `${alias}.${properties.propertyName}` //
+    }
+}
+
+function generatePredicateCondition(
```

Comment:
same, alternatively let's move all filter helper functions already to it's own file

./paginate/filter.ts

then it's okay to keep the broader names